src/ucp/api/ucp.h

```diff
@@ -3135,6 +3108,186 @@ ucs_status_ptr_t ucp_worker_flush_nb(ucp_worker_h worker, unsigned flags,
+ * UCP active message function
+ */
+typedef struct ucp_am_params {
+  uint64_t field_mask;
+  uint64_t flags;
+
+  size_t   iovec_size; /* Number of elements that the ucp library should */
+                       /* allocate in the 'iovec 'for the callback to */
+                       /* fill in. */
+} ucp_am_params_t;
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Completion callback for data arrivals.
+ *
+ * This callback routine is invoked whenever the @ref ucp_tag_send_nb
+ * The ucp library will drive a function on the arrival of each fragment
+ * of data from the initiator.
+ *
+ * @param [out} target    Where the data should be placed. NULL if no copying
+ *                        needed.
+ * @param [in}  source    Where the data currently is
+ * @param [in]  bytes     How much data has just arrived
+ * @param [in]  cookie    Data cookie from the ucp_am_recv structure
+ */
+typedef void (*ucp_am_data_function_t)(void *target, void *source,
+                                       size_t bytes, void *cookie);
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Completion callback for transfer from initiator complete.
+ *
+ * The ucp library will drive a function when the transfer from the
+ * initiator is complete.
+ *
+ * @param [in}  arg             Request structure
+ * @param [in}  cookie          Cookie from the ucp_am_recv structure
+ * @param [in]  iovec           Description of the received data
+ * @param [in]  iovec_length    Number of elements in iovec
+ */
+typedef ucs_status_t (*ucp_am_local_function_t)(void *arg, void *cookie,
+                                                ucp_dt_iov_t *iovec,
+											 size_t iovec_length);
+/**
+ * @brief UCP active message rendezvous redeive parameters field mask.
+ *
+ * The enumeration allows specifying which fields in @ref ucp_am_recv are
+ * present. It is used to enable backward compatibility support.
+ */
+enum ucp_am_params_recv {
+  UCP_AM_PARAMS_RECV_FIELD_FLAGS      = UCS_BIT(0) /* Flags field */
+};
+/*
+ * Structure for communication between the active message
+ * callback and the ucp library. The ucp library sets up 'iovec_max_length'
+ * to indicate the length of the iovec in this structure; the remaining
+ * fields are set by the callback to indicate what should happen
+ * with the remaining data from the initiator. In the initial implementation,
+ * iovec_max_length is always 1.
+ */
+typedef struct ucp_am_recv {
+	uint64_t                field_mask;
+	uint64_t                flags;
+    /* Function to be driven when data transfer is complete */
+    ucp_am_local_function_t local_fn;
+    /* Argument to be passed to local_fn */
+    void                   *cookie;
+    /* Function to be driven when each fragment of data transfer is complete
+     * In the initial implementation, all data is transferred in one fragment,
+     * so the data_fn is called once just before the local_fn .
+     * This implementation is to ease porting of applicaitons currently
```

Comment:
We are designing generic API that has to target multiple applications. Easy porting is nice to have but not an explicit goal. I would remove the last sentence.

Reply:
OK, will do.